Freshbooks integration

ra-ver opened the pull request:
- Created new invoicing service: invoice-api. This interacts with Freshbooks to create, view and email Invoices
- Added new api end points in billing-api that act as interface for invoice-api end points for all the clients
- Other code changes in billing-api to make it to enable support for invoicing

andricDu commented on invoice-api/src/service/freshbooks.ts:
```diff
@@ -0,0 +1,503 @@
+
+    }
+
+    // assumes date is : null or in 'YYYY-MM-DD' format
```
block comments on top of functions

andricDu commented on invoice-api/src/service/freshbooks.ts:
```diff
@@ -0,0 +1,503 @@
+
+    /*
+    Steps to send Invoice:
+        1. Get bearer token using the refresh token
```
👍  great comment!

andricDu commented on invoice-api/src/service/freshbooks.ts:
```diff
@@ -0,0 +1,503 @@
+        return axios.get(`${ this.apiConfig.api }/accounting/account/${ this.apiConfig.account_id }/users/clients?include[]=recent_contacts&search[email_like]=${ email }&page=${ pageNumber }&per_page=100`,
+            {headers: this.headers, httpsAgent: this.agent })
+            .then( response => {
+                console.log(response.data);
```
you might want to remove the logging. Could be pretty verbose.

andricDu commented on invoice-api/src/service/freshbooks.ts:
```diff
@@ -0,0 +1,503 @@
+        userEmail = (userEmail == null || userEmail == "") ?  "":userEmail;
+        let pageCount = 1; let pageIdx = 1;
+        let allCustomersData = [];
+        while(pageCount  >= pageIdx){
```
at first glance the while loop takes a second to understand. Another issue is that it has to make each request sequentially, waiting for the next one to come back before moving forward. 

A better way would be to just fire off all the requests you'll need in parallel, and maintain an array of promises which you can just `await` on with `Promise.all()`

Something like this (my syntax might be wrong):

```javascript
    private async getAllCustomersParallel(userEmail:string): Promise<any>{
        userEmail = (userEmail == null || userEmail == "") ?  "" : userEmail;
        let promises = [];
        let pagedCustomerResults = await this.getAllCustomersPaged(1, userEmail);
        
        for (let i=1; i<pagedCustomerResults.pages; i++) {
            promises.push(this.getAllCustomersPaged(i, userEmail));
        }
        
        let responses = await Promise.all(promises);
        return responses.reduce((a,b) => a.concat(b));
    }
```

cheapsteak replied:
give this a shot

```js
private async getAllCustomersParallel(userEmail:string): Promise<any>{
    let customers = await this.getAllCustomersPaged(1, userEmail || "");
    if (customers.pages > 1) {
      customerResults = customerResults.concat(
        await Promise.all(_.range(1, customers.pages).map(i =>
          this.getAllCustomersPaged(i, userEmail || "")))
      )
    };
    return customers;
}
```

andricDu commented on invoice-api/src/service/freshbooks.ts:
```diff
@@ -0,0 +1,503 @@
+        return axios.get(`${ this.apiConfig.api }/accounting/account/${ this.apiConfig.account_id }/invoices/invoices?search[invoice_number]=${invoiceNumber}`,
+            {headers: this.headers, httpsAgent: this.agent })
+            .then( response => {
+                console.log(response.data);
```
might be too verbose to log.

andricDu commented on invoice-api/src/service/freshbooks.ts:
```diff
@@ -0,0 +1,503 @@
+        return axios.get(`${ this.apiConfig.api }/accounting/account/${ this.apiConfig.account_id }/users/clients?search[userid]=${customerId}&include[]=recent_contacts`,
+            {headers: this.headers, httpsAgent: this.agent })
+            .then( response => {
+                console.log(response.data);
```
another one here, will probably end up being too verbose. Also without timestamps reducing the usefulness.

andricDu commented on invoice-api/src/service/freshbooks.ts:
```diff
@@ -0,0 +1,503 @@
+        return axios.get(`${ this.apiConfig.api }/accounting/account/${ this.apiConfig.account_id }/invoices/invoices?${ searchStr }&page=${ pageNumber }&include[]=lines&per_page=100`,
+            {headers: this.headers, httpsAgent: this.agent })
+            .then( response => {
+                console.log(response.data);
```
another log entry we can probably do away with.

andricDu commented on invoice-api/src/server.ts:
```diff
@@ -0,0 +1,161 @@
+        });
+    });
+
+
```
you can probably get rid of one or two of these newlines.

andricDu commented on billing-invoice/src/service/invoice.ts:
```diff
@@ -0,0 +1,97 @@
+        let dateText = currentDate.toISOString().slice(0,10);
+        return axios.get(`${ this.apiRootURL }/getAllInvoices?date=${ dateText }`)
+            .then( response => {
+                //console.log(response.data);
```
delete this.

cheapsteak commented:
If we want to keep the ability to log around, but only display it when required, would recommend giving this library a look

https://www.npmjs.com/package/debug

Currently the most download package on npm
http://npmcharts.com/compare/lodash,request,async,chalk,express,bluebird,commander,underscore,debug

cheapsteak commented on invoice-api/src/service/freshbooks.ts:
```diff
@@ -0,0 +1,503 @@
+                'discount' : item.discount_value,
+                'total' : item.amount.amount
+            });
+        });
```
Small nitpick, would favor maps over initialize, iterate, and push

```js
let flattenedInovicesJson = allInvoicesForThisPeriod.map(item => (item) => ({
      'current_organization' : item.current_organization,
      'date' : item.create_date,
      'invoice_number' : item.invoice_number,
      'payment_status' : item.payment_status,
      'invoice_status' : item.v3_status,
      'cpu_cost' : this.getLineItemValue("CPU",item.lines),
      'image_cost' : this.getLineItemValue("Image",item.lines),
      'volume_cost' : this.getLineItemValue("Volume",item.lines),
      'discount' : item.discount_value,
      'total' : item.amount.amount
  })
);
```

cheapsteak commented on invoice-api/src/service/freshbooks.ts:
```diff
@@ -0,0 +1,503 @@
+        let minDate = null;
+        let maxDate = null;
+        minDate = date;
+        maxDate = new Date();
```
```diff
- let minDate = null;
- let maxDate = null;
- minDate = date;
- maxDate = new Date();
+ let minDate = date;
+ let maxDate = new Date();
```

cheapsteak commented on invoice-api/src/service/freshbooks.ts:
```diff
@@ -0,0 +1,503 @@
+        let lowerCasedEmail = email.toLowerCase();
+        if(customerInfo.email != ''){
+            if(customerInfo.email.toLowerCase() == lowerCasedEmail) return true;
+        }
```
Can this be

```js
if (customerInfo.email && customerInfo.email.toLowerCase() == lowerCasedEmail) {
 return true;
}
```
?

cheapsteak commented on invoice-api/src/service/freshbooks.ts:
```diff
@@ -0,0 +1,503 @@
+            }
+        } // contacts iteration ends here
+        return false;
+    }
```
Can this be

```js
return !!_.find(
  customerInfo.recent_contacts,
  contact => contact.email && contact.email.toLowerCase() == lowerCasedEmail
);
```
?

cheapsteak replied:
Actually I think the entire function can be reduced to this:

```js
private validEmailForCustomer(email:string, customerInfo:any): boolean{
  return _.every([
    email !== '',
    customerInfo.email.toLowerCase() == lowerCasedEmail,
    !!_.find(
      customerInfo.recent_contacts,
      contact => contact.email && contact.email.toLowerCase() == lowerCasedEmail
    )
  ]);
}
```

ra-ver commented:
Addressed all PR comments and added features/changes requested by Finance and others. Some of Sid's changes are merged in this branch as we were testing it earlier on my machine. However; his latest changes are not integrated in this PR. He raised a separate PR for those changes.
